### PR TITLE
feat(PRIV-1914): replace op-deployer with base/contracts forge scripts in devnet

### PR DESCRIPTION
**Pull request by jowparks**
### Summary
TLDR:
`CONTRACTS_COMMIT=abc123 just devnet up`

Also adds SP1 dependency check for `just devnet up` was failing for me when building locally.

Replace the hardcoded op-deployer binary in the devnet with direct `forge script` calls to [base/contracts](https://github.com/base/contracts), enabling deployment of custom contracts instead of the OP Stack contracts embedded in op-deployer. This is a prerequisite for the privacy appchain work which requires modified contracts.

### Changes

**Docker infrastructure:**
- `Dockerfile.devnet`: Replace op-deployer download with a contracts-builder stage that clones base/contracts at a pinned commit, installs Solidity dependencies, and compiles with forge. Final image uses foundry base (Ubuntu) with `USER root` for runtime tool installation.
- `docker-compose.yml`: Add `patch-genesis-hash` init service that queries the L2 EL node for the real genesis block hash and patches `rollup.json` before consensus nodes start.

**New scripts:**
- `deploy-config.json.template`: Parameterized deploy config for base/contracts, filled via `envsubst` from devnet-env variables.
- `extract-artifacts.sh`: Extracts `l1-addresses.json` from forge deployment output and assembles `genesis.json` from L2Genesis state dump.
- `assemble-rollup-config.sh`: Builds `rollup.json` from L1 RPC state, deployed contract addresses, on-chain SystemConfig values, and protocol parameters. Carefully matches the `RollupConfig` struct which uses `deny_unknown_fields`.
- `L2GenesisDevnet.s.sol`: Thin Solidity wrapper that reads deploy-config + L1 addresses, constructs the `L2Genesis.Input` struct, runs L2Genesis, and dumps state via `vm.dumpState()`.
- `patch-genesis-hash.sh`: Waits for L2 EL to boot, queries genesis block hash, patches rollup.json and rollup-conductor.json.

**Rewritten:**
- `setup-l2.sh`: 5-step pipeline replacing op-deployer: generate deploy-config → forge script SystemDeploy → L2GenesisDevnet → extract artifacts → assemble rollup config. Post-processing (Azul/Beryl activation timestamps, activation admin) preserved unchanged.

**Rust deployer updates:**
- `images.rs`: Remove `OP_DEPLOYER_IMAGE` constant.
- `base_deployer.rs`: Update `DeployerContainer` to use `devnet-setup:local` image with `setup-l2.sh`.
- `deploy_config.rs`: Renamed from `l2_intent.rs`, generates deploy-config JSON instead of op-deployer intent TOML.
- `artifacts.rs`: Add rollup config deserialization validation tests (positive + deny_unknown_fields negative test).

### Testing

- `just devnet up` boots successfully with L2 producing blocks at 2-second cadence
- L1 contracts verified deployed: OptimismPortalProxy, SystemConfigProxy, L1StandardBridgeProxy
- Rollup config deserialization unit tests added (validates against `RollupConfig` with `deny_unknown_fields`)
- Shell scripts pass `shellcheck`

### Known Issues
- L2 transaction test returns "sender is not an EOA" — genesis allocs may set code at Anvil test account addresses, needs investigation in L2Genesis config
- HA sequencer CLs crash — may need additional fields stripped from rollup-conductor.json
- ZK prover build requires pre-built SP1 ELF artifacts (pre-existing, unrelated)

type=routine
risk=low
impact=sev5

### Notes for Reviewers
- The `CONTRACTS_COMMIT` ARG in Dockerfile.devnet is pinned to `c8f0238ec570a446e01b6290bd2413ef7704fafd` (current HEAD of base/contracts main). Update as needed.
- `FOUNDRY_SCRIPT_EXECUTION_PROTECTION=false` is set on forge script calls because base/contracts `SystemDeploy.s.sol` uses `address(this)` which forge v1.7+ blocks by default.
- The `patch-genesis-hash` docker-compose service is an architectural choice: the L2 genesis hash can only be computed after reth initializes from genesis.json, but rollup.json (which needs the hash) must exist before consensus nodes start. The init service bridges this gap.
- base/contracts `L2Genesis.s.sol` has a different API than the OP Stack version — it takes an `Input` struct rather than reading from deploy-config. The `L2GenesisDevnet.s.sol` wrapper bridges this.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: etc/docker/Dockerfile.devnet [hunk 1]
```diff
@@ -1,22 +1,13 @@
-FROM public.ecr.aws/docker/library/golang:1.26-alpine AS builder
+FROM public.ecr.aws/docker/library/golang:1.26 AS builder
 
 # Install build dependencies for CGO (required for BLS library)
```
Comment (jowparks) [inline etc/docker/Dockerfile.devnet:3]: required for forge image to work below.

File: etc/docker/Dockerfile.devnet [hunk 2]
```diff
@@ -33,14 +24,48 @@ RUN git clone https://github.com/protolambda/eth2-val-tools.git --branch "${ETH2
     test "$(git rev-parse HEAD)" = "${ETH2_VAL_TOOLS_COMMIT}"
 RUN go build -o /go/bin/eth2-val-tools .
 
-FROM public.ecr.aws/docker/library/alpine:3.21.3
+# Build contracts using Foundry
+# The foundry image is Ubuntu-based and multi-arch (amd64/arm64)
+FROM ghcr.io/foundry-rs/foundry:latest AS contracts-builder
```
Comment (jowparks) [inline etc/docker/Dockerfile.devnet:29]: Had to change from alpine to get foundry commands to work

**Comment by jowparks** [comment]
Superseded by #3120 which targets main directly. The `privacy` branch is long-lived and should not be merged.
